### PR TITLE
Make injector generic

### PR DESCRIPTION
**Pull request by mchandramouli**
This PR makes sidecar injector generic. Instead of just injecting haystack-agent, it can inject any configured sidecar by its name -and- injects multiple sidecars if the pod is configured with a list of sidecars to be injected

### CODE REVIEW

File: webhook/mutator.go [hunk 1]
```diff
@@ -125,55 +125,79 @@ func unMarshall(req *v1beta1.AdmissionRequest) (corev1.Pod, error) {
...
-	}
+func createPatch(pod *corev1.Pod, sideCarNames []string, sideCars map[string]*SideCar, annotations map[string]string) ([]byte, error) {
+
+	var patch []patchOperation
+	var containers       []corev1.Container
+	var volumes          []corev1.Volume
+	var imagePullSecrets []corev1.LocalObjectReference
+	count := 0
+
+	for _, name := range sideCarNames {
```
Comment (ayansen) [inline webhook/mutator.go]: Will it only be able to append containers, volumes, imagePullSecrets ?

Comment (mchandramouli) [reply webhook/mutator.go]: @ayansen that is all we are having in `webhook.Sidecar` struct.

Comment (mchandramouli) [reply webhook/mutator.go]: Looking at Pod spec, these covers most of what is needed to inject a sidecar

Comment (ayansen) [reply webhook/mutator.go]: sounds good. I was thinking about initContainers but I guess most sidecars wouldn't need one. This looks good for now

**Comment by ayansen** [comment]
The changes look good. Will test it out locally once and merge it in

File: webhook/mutator.go (hunk 1 above)
Comment (jcchavezs) [inline webhook/mutator.go]: I am curious, can't you just do:
`for name, sideCar := range sideCars {...}`

Comment (ayansen) [reply webhook/mutator.go]: @jcchavezs I'm seeing the pod can specify which sidecars to inject by passing an array of  sidecar names in an annotation and thats why I think the loop isn't iterating over the the entire sidecar map.  SidecarNames based on my understanding is a subset of the the sidecar map

**Comment by jcchavezs** [comment]
:+1:

Den tir. 5. mar. 2019, 11:16 skrev Ayan Sen <notifications@github.com>:

**Comment by jcchavezs** [comment]
Just a readability tip: calling them availableSideCars an enabled sidecars
would help to understand this.

Den tir. 5. mar. 2019, 11:16 skrev Ayan Sen <notifications@github.com>:

**Pull request merged**